experimental/sql/src/alg_to_ssa.py

```diff
@@ -0,0 +1,163 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import RewritePattern, GreedyRewritePatternApplier, PatternRewriteWalker, PatternRewriter, op_type_rewrite_pattern
+
+import dialects.rel_alg as RelAlg
+import dialects.rel_ssa as RelSSA
+
+# This file contains the rewrite infrastructure to translate the relational
+# algebra dialect to the relational SSA dialect. The current design has a parent
+# class `RelAlgRewriter` that contains functions used for several `Rewriter`s.
+# All other `Rewriter`s inherit from that class. To introduce SSA, the passes
+# insert the children (and the their whole block), that should become operands,
+# before the matched op, then use the `added_operations_before` functionality of
+# the rewriter to link the operands to the new operation.
+
+
+@dataclass
+class RelAlgRewriter(RewritePattern):
+
+  def convert_datatype(self, type_: RelAlg.DataType) -> RelSSA.DataType:
```

Comment:
This is quite strange to me that a RewritePattern allows type conversion.
There are non-obvious caveats on the MLIR side re conversion infra, type converter, rollbacks, implicit type conversions introduced by the infra, verification and how all these mix with block arguments and returns.

Is there a good doc / rationale for this? Maybe loop in Tobias and/or Matthieu ?

Reply:
This actually goes beyond my understanding of MLIR. Maybe @math-fehr or @tobiasgrosser can help?

Reply:
So here, this is not really a type conversion like in MLIR AFAIU.
This is just a method that will convert a type to another, and this method is used in the rewrite pattern.
Like in here:
```
  def match_and_rewrite(self, op: RelAlg.Literal, rewriter: PatternRewriter):
    new_op = RelSSA.Literal.get(op.val, self.convert_datatype(op.type))
    rewriter.replace_matched_op([new_op, RelSSA.Yield.get([new_op])])
```

xDSL does not have support for type conversion AFAIK. Currently, we are just writing rewrite patterns that will happen to convert types in blocks contained in the operations.